### DISCUSSION TITLE
sync ostream buffer after whole json write finished

### DISCUSSION THREAD

**Discussion by Shadasviar** [nlohmann/json#2752] [Ideas]
I have a custom socket implementation, which inherits from` streambuf` and  `iostream`, so I can use it with `nlohmann::json` like below:

```
Socket s("/tmp/out");
nlohmann::json json = ... /*some valid json here*/
s << json;
s.flush();
```

As you can see from example code, I always need to call streambuf `flush` method after json was written to the socket, because nlohmann library doesn't call it inside.

Of course, I can implement `putc` function of my buffer such way, so it would call `sync` and perform write system call for every char/only for end of line character, but I think that it would be good if nlohmann library would call `sync` function on output streambuf it use after json was put to buffer, because library knows when json ends. Such behavior could save a lot of write system calls and user should not to call `flush` manually.

I would want to discuss, if adding syncing buffer to the nlohmann library is a good or bad idea, does it have some drawbacks etc.

**Comment by nlohmann** [comment]
I understand the issue, but I am not sure whether the library should decide when to flush - I think this is something an application can better decide. In your case, you could create a function that wraps the streaming/flushing. In a general case, the use case may dictate different strategies - maybe it's perfectly fine to never flush.
